README.md
![Logo](/logo.png)

# React-mapbox-gl | [Documentation](docs/API.md) | [Demos](http://alex3165.github.io/react-mapbox-gl/demos)

[![Build Status](https://travis-ci.org/alex3165/react-mapbox-gl.svg?branch=master)](https://travis-ci.org/alex3165/react-mapbox-gl)
[![npm version](https://img.shields.io/npm/v/react-mapbox-gl.svg?style=flat)](https://www.npmjs.com/package/react-mapbox-gl)
[![npm downloads](https://img.shields.io/npm/dm/react-mapbox-gl.svg)](https://www.npmjs.com/package/react-mapbox-gl)
[![TypeScript](https://badges.frapsoft.com/typescript/code/typescript.svg?v=101)](https://github.com/ellerbrock/typescript-badges/)
<br/>

 #### React wrapper for [mapbox-gl-js](https://www.mapbox.com/mapbox-gl-js/api/).
<br/><br/>
![London cycle example gif](docs/london-cycle-example.gif "London cycle example gif")

## Components

### Proxy components (proxy between React and Mapbox API)
- ReactMapboxGL
- Layer & Feature
  - property `symbol` displays a mapbox symbol.
  - property `line` displays a lineString.
  - property `fill` displays a polygon.
  - property `circle` displays a mapbox circle.
  - property `raster` displays a mapbox raster tiles.
  - property `fill-extrusion` displays a layer with extruded buildings.
  - property `background` displays a mapbox background layer.
  - property `heatmap` displays a mapbox heatmap layer.
- Source
- GeoJSONLayer

### DOM components (normal React components)
- ZoomControl
- ScaleControl
- RotationControl
- Marker (Projected component)
- Popup (Projected component)
- Cluster

## Getting Started

```
npm install react-mapbox-gl mapbox-gl --save
```

Example:

```jsx
// ES6
import ReactMapboxGl, { Layer, Feature } from "react-mapbox-gl";

// ES5
var ReactMapboxGl = require("react-mapbox-gl");
var Layer = ReactMapboxGl.FeatureLayer;
var Feature = ReactMapboxGl.Feature;

const Map = ReactMapboxGl({
  accessToken: "pk.eyJ1IjoiZmFicmljOCIsImEiOiJjaWc5aTV1ZzUwMDJwdzJrb2w0dXRmc2d0In0.p6GGlfyV-WksaDV_KdN27A"
});

// in render()
<Map
  style="mapbox://styles/mapbox/streets-v9"
  containerStyle={{
    height: "100vh",
    width: "100vw"
  }}>
    <Layer
      type="symbol"
      id="marker"
      layout={{ "icon-image": "marker-15" }}>
      <Feature coordinates={[-0.481747846041145, 51.3233379650232]}/>
    </Layer>
</Map>
```

## Why are `zoom`, `bearing` and `pitch` Arrays ?
If those properties changed at the mapbox-gl-js level and you don't update the value kept in your state, it will be unsynced with the current viewport. At some point you might want to update the viewport value (zoom, pitch or bearing) with the ones in your state but using value equality is not enough. Taking zoom as example, you will still have the unsynced zoom value therefore we can't tell if you want to update the prop or not. In order to explicitly update the current viewport values you can instead break the references of those props and reliably update the current viewport with the one you have in your state to be synced again.

## [Version 3.0 Documentation](docs/API.md)
## [Version 2.0 Documentation](https://github.com/alex3165/react-mapbox-gl/blob/v2-archive/docs/API.md)

## Contributions
Please try to reproduce your problem with the [boilerplate](https://github.com/alex3165/react-mapbox-gl-debug) or [this webpackbin template](https://www.webpackbin.com/bins/-L2MXLm-gCwvABQZfSnf) before posting an issue.

## Built with react-mapbox-gl
[monumap.org](https://monumap.org/)

## mapbox-gl-draw compatibility
Try [react-mapbox-gl-draw](https://github.com/amaurymartiny/react-mapbox-gl-draw)

## Looking for an Angular alternative?
Try [ngx-mapbox-gl](https://github.com/Wykks/ngx-mapbox-gl)


CHANGELOG.md
## Above 2.0.0

See [release notes](https://github.com/alex3165/react-mapbox-gl/releases)

## 2.0.* (June 2017)
### Breaking changes
- `ReactMapboxGl` is now a factory that return a React Component, see documentation on how to use it

### Feature
- `Feature` component has a `draggable` property
- The map component returned by the factory accept all the events
- Listen events only on the concerned `Layer`, should improve performances

### Bug fixes
- `Cluster` component bugs have been fixed, it is tested and should behave as expected
- Fix issue when using `text-align: center` on a parent tag of the map
- Fix wrong unmounting of the map
- Fix `Layer` and `GeoJsonLayer` when changing the style of the map

### Others
- Improvement of the documentation
- Brand new website for the examples with more examples
- Add more tests
- Prettify all the source files
- Update dependencies

## 1.4.0 (March 14 2017)

- Breaking change : `onHover` is now `onMouseEnter`, `onEndHover` is now `onMouseLeave`
- Breaking change: `GEOJSONLayer` is not displaying all the layers in your geojson source, you have to specify which source to display by passing the correct props, see [note](https://github.com/alex3165/react-mapbox-gl/blob/master/docs/API.md#geojsonlayer)
- Fix issue when unmounting all the features `setData` is not called and the features doesn't disappear. [PR](https://github.com/alex3165/react-mapbox-gl/pull/152)

## 1.0.0 (February 13 2017)

- All the source code is now written in Typescript, you can now get the type definitions for each component in a Typescript environment.

## 0.19.0 (December 12 2016)

- Implement the `Popup` and the `Marker` Components using React. These components now keep track of the context.

## 0.17.2 (November 30 2016)

### Features

- Added `zIndex` and `onClick` properties to Marker

## 0.17.1 (November 29 2016)

- Added a RasterLayer for both WMS and mapbox raster

## 0.17.0 (November 28 2016)

:warning: BREAKING :warning:
- Renamed layer to FeatureLayer to more accurately describe what it actually does

- Upgrade mapbox-gl to v0.28.0
- Added the fill-extrusion option (including example)
- General dep upgrade
- General linting fixes

## 0.15.0 (October 23 2016)

- Upgrade mapbox-gl from v0.21.0 to v0.25.1 and don't build the vendor anymore, react-mapbox-gl is now using mapbox-gl/dist/mapbox-gl.js bundle again
- Improve the examples with a navigation to select one.
- Fix a major issue when removing a layer with a sourceId.

## 0.11.0 (September 8 2016)

- Add GeoJSONLayer component to display a raw geojson

## 0.8.0 (July 5 2016)

- Breaking change, zoom is now wrapped in an array so the map component can check for a reference equality of the value.

## 0.7.1 (June 16 2016)

- Update API documentation
- A change of the style geojson passed down to the map component will update the style of the map itself using mapbox `setStyle`
- Get rid of `lodash` for internal methods and smaller packages


## 0.7.0 (June 15 2016)

- Add `layerOptions` property to Layer component
- Layer can use external source : #22
- Add Layer with external source to all-shape example
- `onStyleLoad` callback is now called before childrens components are rendered
- Update mapbox-gl to version 0.20.0
- Add properties property to Feature component


example/src/all-shapes.js
/* eslint-disable no-console, no-param-reassign */

import React, { Component } from 'react';
import ReactMapboxGl, { FeatureLayer, Feature, ScaleControl, ZoomControl, Marker } from 'react-mapbox-gl'; // eslint-disable-line
import route from './route.json';
import config from './config.json';
import style from './style.json';

const { accessToken } = config;

const containerStyle = {
  height: '100vh',
  width: '100%',
};

const polygonCoords = [[
  [-0.13235092163085938, 51.518250335096376],
  [-0.1174163818359375, 51.52433860667918],
  [-0.10591506958007812, 51.51974577545329],
  [-0.10831832885742188, 51.51429786349477],
  [-0.12531280517578122, 51.51429786349477],
  [-0.13200759887695312, 51.517823057404094],
]];

const multiPolygonCoords = [
  [[
    [-0.18235092163085938, 51.518250335096376],
    [-0.1674163818359375, 51.52433860667918],
    [-0.15591506958007812, 51.51974577545329],
    [-0.15831832885742188, 51.51429786349477],
    [-0.17531280517578122, 51.51429786349477],
    [-0.18200759887695312, 51.517823057404094],
  ]],
  [[
    [-0.18235092163085938, 51.538250335096376],
    [-0.1674163818359375, 51.54433860667918],
    [-0.15591506958007812, 51.53974577545329],
    [-0.15831832885742188, 51.53429786349477],
    [-0.17531280517578122, 51.53429786349477],
    [-0.18200759887695312, 51.537823057404094],
  ]],
];

const markerCoord = [
  -0.2416815,
  51.5285582,
];

const mappedRoute = route.points.map(point => [point.lat, point.lng]);

const markerContainer = document.createElement('div');
markerContainer.style.width = '400px';
markerContainer.style.position = 'absolute';

export default class AllShapes extends Component {

  state = {
    popup: null,
    center: [0.2174037, 51.6476704],
    circleRadius: 30,
    routeIndex: 0,
  };

  componentWillMount() {
    setTimeout(() => {
      this.setState({
        center: [-0.120736, 51.5118219],
        circleRadius: 10,
      });
    }, 6000);

    setInterval(() => {
      this.setState({
        routeIndex: this.state.routeIndex + 1,
      });
    }, 8000);
  }

  _onClickMarker = ({ feature }) => {
    this.setState({
      center: feature.geometry.coordinates,
    });
  };

  _onClickMap(map) {
    console.log('Clicked on the map : ', map);
  }

  _onStyleLoad(map) {
    console.log('Style loaded: ', map);
  }

  _onHover({ map }) {
    map.getCanvas().style.cursor = 'pointer';
  }

  _onZoom(map) {
    console.log('Zoom level changed to ', map.getZoom());
  }

  _onEndHover({ map }) {
    map.getCanvas().style.cursor = '';
  }

  _polygonClicked = ({ feature }) => {
    console.log('Polygon clicked', feature.geometry.coordinates);
  };

  render() {
    return (
      <ReactMapboxGl
        style={style}
        onClick={this._onClickMap}
        onZoom={this._onZoom}
        onStyleLoad={this._onStyleLoad}
        accessToken={accessToken}
        center={this.state.center}
        movingMethod="jumpTo"
        containerStyle={containerStyle}>
        <ScaleControl/>
        <ZoomControl/>
        <FeatureLayer
          type="symbol"
          layout={{ 'icon-image': 'harbor-15' }}>
          <Feature
            coordinates={markerCoord}
            onHover={this._onHover}
            onEndHover={this._onEndHover}
            onClick={this._onClickMarker}/>
        </FeatureLayer>

        <FeatureLayer
          id="mapbox-route-example"
          type="line"
          sourceId="route"
          layout={{
            'line-join': 'round',
            'line-cap': 'round',
          }}
          paint={{
            'line-color': '#888',
            'line-width': 8,
          }}/>

        <FeatureLayer
          type="line"
          layout={{ 'line-cap': 'round', 'line-join': 'round' }}
          paint={{ 'line-color': '#4790E5', 'line-width': 12 }}>
          <Feature coordinates={mappedRoute}/>
        </FeatureLayer>

        <FeatureLayer
          type="circle"
          paint={{ 'circle-radius': this.state.circleRadius, 'circle-color': '#E54E52', 'circle-opacity': 0.8 }}>
          <Feature coordinates={mappedRoute[this.state.routeIndex]}/>
        </FeatureLayer>

        <Marker
          container={markerContainer}
          coordinates={markerCoord}
          onClick={e => console.log('Clicked the marker', e)}
        >
          <h1>
            TEST
          </h1>
        </Marker>

        <FeatureLayer
          type="fill"
          paint={{ 'fill-color': '#6F788A', 'fill-opacity': 0.7 }}>
          <Feature
            onClick={this._polygonClicked}
            coordinates={polygonCoords}/>
        </FeatureLayer>

        <FeatureLayer
          type="fill"
          paint={{ 'fill-color': '#3bb2d0', 'fill-opacity': 0.5 }}>
          <Feature
            onClick={this._polygonClicked}
            coordinates={multiPolygonCoords}/>
        </FeatureLayer>

      </ReactMapboxGl>
    );
  }
}
